docs/Modern-PersistedQueries.md

```diff
@@ -0,0 +1,149 @@
+query string. This saves upload bytes from the client to the server.
+
+* the server can now whitelist queries which improves security by restricting the operations 
+that can be run from the client.
+
+## Usage on the client
+
+### The `--persist` flag
+In your `npm` script in `package.json`, run the relay compiler using the `--persist` flag:
+
+```js
+"scripts": {
+  "relay": "relay-compiler --src ./src --schema ./schema.graphql --persist"
+}
+```
+
+The `--persist` flag does 3 things:
+
+1. It converts all query and mutation operation texts to md5 hashes.
+
+    For example without `--persist`, a generated `ConcreteRequest` might look like below:
+
+    ```js
+    const node/*: ConcreteRequest*/ = (function(){
+    //... excluded for brevity
+    return {
+      "kind": "Request",
+      "operationKind": "query",
+      "name": "TodoItemRefetchQuery",
+      "id": null, // NOTE: id is null
+      "text": "query TodoItemRefetchQuery(\n  $itemID: ID!\n) {\n  node(id: $itemID) {\n    ...TodoItem_item_2FOrhs\n  }\n}\n\nfragment TodoItem_item_2FOrhs on Todo {\n    text\n    isComplete\n}\n",
+      //... excluded for brevity
+    };
+    })();
+    ```
+    
+    With `--persist` this becomes:
+    
+    ```js
+    const node/*: ConcreteRequest*/ = (function(){
+    //... excluded for brevity
+    return {
+      "kind": "Request",
+      "operationKind": "query",
+      "name": "TodoItemRefetchQuery",
+      "id": "3be4abb81fa595e25eb725b2c6a87508", // NOTE: id is now an md5 hash of the query text
+      "text": null, // NOTE: text is null now
+      //... excluded for brevity
+    };
+    })();
+    ```
+
+2. It generates a matching `.graphql.json` file containing a map of the id and the operation text in the same `__generated__` 
+directory as the `.graphql.js` file. In the example above, the `__generated__` directory will have these files:
+
+    * `./__generated__/TodoItemRefetchQuery.graphql.js`
+    * `./__generated__/TodoItemRefetchQuery.graphql.json`
+    
+    The `.graphql.json` file looks something like this:
+    
+    ```json
+    {
+      "3be4abb81fa595e25eb725b2c6a87508": "query TodoItemRefetchQuery(\n  $itemID: ID!\n) {\n  node(id: $itemID) {\n    ...TodoItem_item_2FOrhs\n  }\n}\n\nfragment TodoItem_item_2FOrhs on Todo {\n    text\n    isComplete\n}\n"
+    }
+    ```
+
+3. It also generates a complete query map file at `[your_src_dir]/queryMap.graphql.json`. This file contains all the query ids 
+and their operation texts.
+
+### Network layer changes
+You'll need to modify your network layer fetch implementation to pass a queryId parameter in the POST body instead of a query parameter:
+
+```js
+function fetchQuery(operation, variables,) {
+  return fetch('/graphql', {
+    method: 'POST',
+    headers: {
+      'content-type': 'application/json'
+    },
+    body: JSON.stringify({
+      queryId: operation.id, // NOTE: pass md5 hash to the server
+      // query: operation.text, // this is now obsolete because text is null 
```

Comment:
It should be noted there is a 3rd, but much more complex, way to get the queries to the server. If all you care about is performance, then you can optimistically send the ID to the server. If the server responds that it doesn't know the ID, then you can send the whole query text, which the server can then cache. The next time you send the query ID to the server, it will have it. This is how [apollo-link-persisted-queries](https://www.npmjs.com/package/apollo-link-persisted-queries) works. I don't think we need to write an implementation of this, but I think we should note in the docs here that it is possible.

Reply:
That's an interesting approach. It seems to conflict with whitelisting though? You can't optimistically send queries if you are also whitelisting queries. Perhaps I should include this point in the docs as well. Will update.

Reply:
Yeah, that's true, but it is an option of all you're looking to do is optimization. I lot of people won't be able to whitelist because they allow 3rd party queries, but this option allows them (or people who don't want to/can't update a central repository) to still cache queries.

Reply:
Gotcha. Will include in the docs. Thanks for the explanation!

Reply:
Done @MatthewHerbst